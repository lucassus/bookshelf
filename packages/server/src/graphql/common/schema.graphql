scalar ExternalID

interface Resource {
  id: ExternalID!
}

union Anything = Author | Book | User | BookCopy

interface Timestampable {
  createdAt: String!
  updatedAt: String!
}

type Image {
  path: String!
  url: String!
}

extend type Query {
  resource(id: ID!): Resource!
  anything(id: ID!): Anything! @deprecated(reason: "No longer supported. Use 'resources' instead")
}
