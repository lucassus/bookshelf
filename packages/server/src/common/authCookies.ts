import express from "express";

import {
  AUTH_COOKIE_NAME,
  AUTH_TOKEN_EXPIRES_IN_SECONDS,
  Environment
} from "../config";

export const sendAuthCookie = (res: express.Response, authToken: string) =>
  res.cookie(AUTH_COOKIE_NAME, authToken, {
    httpOnly: true,
    secure: process.env.NODE_ENV === Environment.production,
    maxAge: AUTH_TOKEN_EXPIRES_IN_SECONDS * 1000
  });

export const clearAuthCookie = (res: express.Response) =>
  res.clearCookie(AUTH_COOKIE_NAME);
