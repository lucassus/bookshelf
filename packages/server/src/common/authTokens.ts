import jwt from "jsonwebtoken";

import {
  AUTH_TOKEN_EXPIRES_IN_SECONDS,
  AUTH_TOKEN_SECRET_KEY
} from "../config";
import { User } from "../database/entity";

export const getAuthTokenSecretFor = (user: User) =>
  [user.email, user.passwordHash, AUTH_TOKEN_SECRET_KEY].join(".");

export const generateAuthToken = (user: User): string =>
  jwt.sign({ sub: user.id }, getAuthTokenSecretFor(user), {
    expiresIn: AUTH_TOKEN_EXPIRES_IN_SECONDS
  });
