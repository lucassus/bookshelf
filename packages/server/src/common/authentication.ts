import bcrypt from "bcrypt";
import { Request } from "express";
import jwt from "jsonwebtoken";

import { AUTH_TOKEN_EXPIRES_IN, AUTH_TOKEN_SECRET_KEY } from "../config";
import { User } from "../database/entity/User";

type AuthTokenPayload = {
  sub: number;
};

export function hashPassword(password: string): string {
  const salt = bcrypt.genSaltSync(8);
  return bcrypt.hashSync(password, salt);
}

export function isPasswordValid(password: string, hash: string): boolean {
  return bcrypt.compareSync(password, hash);
}

export function generateAuthToken(user: User): string {
  const payload: AuthTokenPayload = {
    sub: user.id
  };

  return jwt.sign(payload, AUTH_TOKEN_SECRET_KEY, {
    expiresIn: AUTH_TOKEN_EXPIRES_IN
  });
}

export function authenticateRequest(req: Request): null | number {
  const { authorization } = req.headers;

  if (!authorization) {
    return null;
  }

  const prefix = "Bearer ";
  if (!authorization.startsWith(prefix)) {
    return null;
  }

  const authToken = authorization.substring(
    prefix.length,
    authorization.length
  );

  const payload = jwt.verify(
    authToken,
    AUTH_TOKEN_SECRET_KEY
  ) as AuthTokenPayload;

  return payload.sub;
}

export function checkAuthentication(currentUser: User | undefined): User {
  if (!currentUser) {
    throw new Error("Unauthorized access! Please log in.");
  }

  return currentUser;
}

export function checkAdminAuthentication(currentUser: User | undefined): User {
  const user = checkAuthentication(currentUser);

  if (!user.isAdmin) {
    throw new Error("Unauthorized access! Please log in as admin.");
  }

  return user;
}
