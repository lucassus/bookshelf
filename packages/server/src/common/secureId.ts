import { SecureId } from "@bookshelf/secure-id";

export const secureId = new SecureId<string>({
  separator: "-"
});
