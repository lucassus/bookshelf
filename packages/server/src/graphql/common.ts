import { Author } from "../database/entity/Author";
import { Avatar } from "../database/entity/Avatar";
import { Book } from "../database/entity/Book";
import { Context } from "../types";
import { Resolver, ResolversTypes } from "./resolvers-types.generated";

// TODO: Figure out how to refactor it with interface
export const image: Resolver<
  ResolversTypes["Image"],
  Avatar | Author | Book,
  Context
> = (rootValue, args, { assetsBaseUrl }) => {
  let path = "/default.jpg";

  if (rootValue instanceof Avatar) {
    path = rootValue.imagePath;
  }

  if (rootValue instanceof Author) {
    path = rootValue.photoPath;
  }

  if (rootValue instanceof Book) {
    path = rootValue.coverPath;
  }

  return {
    path,
    url: assetsBaseUrl + path
  };
};
