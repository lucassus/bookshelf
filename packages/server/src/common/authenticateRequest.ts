import express from "express";
import jwt from "jsonwebtoken";
import { getRepository } from "typeorm";

import { AUTH_COOKIE_NAME } from "../config";
import { User } from "../database/entity";
import { getAuthTokenSecretFor } from "./authTokens";

export async function authenticateRequest(
  req: express.Request
): Promise<undefined | User> {
  const { [AUTH_COOKIE_NAME]: authToken } = req.cookies;
  if (!authToken) {
    return undefined;
  }

  const payload = jwt.decode(authToken);
  if (!(payload && payload.sub)) {
    throw new Error("Invalid token payload");
  }

  const user = await getRepository(User).findOneOrFail({
    id: payload.sub
  });

  return new Promise((resolve, reject) => {
    jwt.verify(authToken, getAuthTokenSecretFor(user), (error) => {
      if (error) {
        reject(error);
      } else {
        resolve(user);
      }
    });
  });
}
