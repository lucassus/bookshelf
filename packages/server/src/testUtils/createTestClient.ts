import { ApolloServer } from "apollo-server-express";
import {
  ApolloServerTestClient,
  createTestClient as createApolloTestClient
} from "apollo-server-testing";

import { buildContext, Context } from "../graphql/context";
import { rootSchema } from "../graphql/rootSchema";

export function createTestClient(
  contextExtra: Partial<Context> = {}
): ApolloServerTestClient {
  const server = new ApolloServer({
    schema: rootSchema,
    context: buildContext(contextExtra)
  });

  return createApolloTestClient(server);
}
