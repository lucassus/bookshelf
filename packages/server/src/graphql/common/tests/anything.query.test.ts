import { gql } from "apollo-server-express";

import { secureId } from "../../../common/secureId";
import { createTestClient } from "../../../testUtils/createTestClient";
import {
  createAuthor,
  createBook,
  createBookCopy,
  createUser
} from "../../../testUtils/factories";

describe("anything query", () => {
  const GetAnythingQuery = gql`
    query($id: ID!) {
      anything(id: $id) {
        __typename

        ... on Author {
          id
          name
          books {
            ...BookFragment
          }
        }

        ...BookFragment

        ...UserFragment

        ... on BookCopy {
          id
          owner {
            ...UserFragment
          }
          borrower {
            ...UserFragment
          }
          book {
            ...BookFragment
          }
        }
      }
    }

    fragment BookFragment on Book {
      id
      title
      description
      favourite
    }

    fragment UserFragment on User {
      id
      name
    }
  `;

  it("fetches Author", async () => {
    // Given
    const author = await createAuthor();
    await createBook({ author });
    await createBook({ author });
    await createBook({ author });

    // When
    const res = await createTestClient().query({
      query: GetAnythingQuery,
      variables: { id: secureId.toExternal(author.id, "Author") }
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchSnapshot();
  });

  it("fetches Book", async () => {
    // Given
    const book = await createBook();

    // When
    const res = await createTestClient().query({
      query: GetAnythingQuery,
      variables: { id: secureId.toExternal(book.id, "Book") }
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchSnapshot();
  });

  it("fetches User", async () => {
    // Given
    const user = await createUser();

    // When
    const res = await createTestClient().query({
      query: GetAnythingQuery,
      variables: { id: secureId.toExternal(user.id, "User") }
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchSnapshot();
  });

  it("fetches BookCopy", async () => {
    // Given
    const borrower = await createUser();
    const bookCopy = await createBookCopy({ borrower });

    // When
    const res = await createTestClient().query({
      query: GetAnythingQuery,
      variables: { id: secureId.toExternal(bookCopy.id, "BookCopy") }
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchSnapshot();
  });
});

it("fetches with aliases", async () => {
  // Given
  const book = await createBook();
  const author = await createAuthor();
  const user = await createUser();

  // When
  const res = await createTestClient().query({
    query: gql`
      query(
        $bookId: ExternalID!
        $authorId: ExternalID!
        $userId: ExternalID!
      ) {
        something: book(id: $bookId) {
          id
          externalId: id
          headline: title
          description
        }

        author(id: $authorId) {
          id
          headline: name
          description: bio
        }

        user(id: $userId) {
          id
          headline: name
          description: info
        }
      }
    `,
    variables: {
      bookId: secureId.toExternal(book.id, "Book"),
      authorId: secureId.toExternal(author.id, "Author"),
      userId: secureId.toExternal(user.id, "User")
    }
  });

  // Then
  expect(res.errors).toBe(undefined);
  expect(res.data).toMatchSnapshot();
});
