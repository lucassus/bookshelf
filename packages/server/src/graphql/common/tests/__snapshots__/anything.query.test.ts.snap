// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anything query fetches Author 1`] = `
Object {
  "__typename": "Author",
  "books": Array [
    Object {
      "description": "Pariatur laboriosam aliquid velit molestiae voluptates exercitationem reiciendis.",
      "favourite": true,
      "id": "My1Cb29r",
      "title": "Dolore Tenetur Voluptatem Nostrum",
    },
    Object {
      "description": "Ipsam exercitationem sequi est minus consequatur minus fugit.",
      "favourite": false,
      "id": "Mi1Cb29r",
      "title": "Qui Adipisci Qui",
    },
    Object {
      "description": "Laudantium molestias ut.",
      "favourite": true,
      "id": "MS1Cb29r",
      "title": "Veniam",
    },
  ],
  "id": "MS1BdXRob3I=",
  "name": "Rosalind Effertz",
}
`;

exports[`anything query fetches Book 1`] = `
Object {
  "__typename": "Book",
  "description": "Laudantium molestias ut.",
  "favourite": true,
  "id": "MS1Cb29r",
  "title": "Veniam",
}
`;

exports[`anything query fetches BookCopy 1`] = `
Object {
  "__typename": "BookCopy",
  "book": Object {
    "description": "Vero aut assumenda.",
    "favourite": true,
    "id": "MS1Cb29r",
    "title": "Minus",
  },
  "borrower": Object {
    "id": "MS1Vc2Vy",
    "name": "Mariana Christiansen",
  },
  "id": "MS1Cb29rQ29weQ==",
  "owner": Object {
    "id": "Mi1Vc2Vy",
    "name": "Eusebio Heidenreich",
  },
}
`;

exports[`anything query fetches User 1`] = `
Object {
  "__typename": "User",
  "id": "MS1Vc2Vy",
  "name": "Mariana Christiansen",
}
`;

exports[`fetches with aliases 1`] = `
Object {
  "author": Object {
    "description": "Ipsam exercitationem sequi est minus consequatur minus fugit.",
    "headline": "Clemens Hyatt",
    "id": "Mi1BdXRob3I=",
  },
  "something": Object {
    "description": "Laudantium molestias ut.",
    "externalId": "MS1Cb29r",
    "headline": "Veniam",
    "id": "MS1Cb29r",
  },
  "user": Object {
    "description": "Voluptates exercitationem reiciendis cupiditate provident earum quaerat architecto qui ad.",
    "headline": "Alba Green",
    "id": "MS1Vc2Vy",
  },
}
`;
